tests/gbs/test_sample.py

```diff
@@ -58,309 +41,97 @@ def test_invalid_adjacency(self, dim):
-
-    @pytest.mark.parametrize("valid_backends", sample.QUANTUM_BACKENDS)
-    def test_pnr_postselect(self, adj, monkeypatch, valid_backends):
-        """Test if function returns samples when photon-number resolving detection is used and
-        postselection. Postselection only returns samples with a total photon number not less
-        than the parameter specified. """
-
-        def dummy_sample_sf(*args, **kwargs):
-            """Dummy ``_sample_sf`` function"""
-            return sample_pnr_postselect
+        p.append(
+            sf.ops.GraphEmbed(adj, mean_photon_per_mode=mean_photon_per_mode), list(range(dim))
+        )
+        p.append(sf.ops.MeasureFock(), list(range(dim)))
 
-        with monkeypatch.context() as m:
-            m.setattr(sample, "_sample_sf", dummy_sample_sf)
-            samples = np.array(
-                sample.sample(
-                    A=adj,
-                    n_mean=1.0,
-                    samples=sample_number,
-                    backend_options={
-                        "backend": valid_backends,
-                        "threshold": False,
-                        "postselect": 1,
-                    },
-                )
-            )
-
-        samples_target = np.array([sample_pnr_postselect[0] for _ in range(sample_number)])
-
-        assert np.allclose(samples, samples_target)
-
-    @pytest.mark.parametrize("valid_backends", sample.QUANTUM_BACKENDS)
-    def test_threshold_postselect(self, adj, monkeypatch, valid_backends):
-        """Test if function returns samples when threshold detection is used and postselection.
-        With threshold detection, all non-zero elements of a sample are set to 1. Postselection
-        only returns samples with a total click number (number of 1's) not less than the
-        parameter specified. """
-
-        def dummy_sample_sf(*args, **kwargs):
-            """Dummy ``_sample_sf`` function"""
-            return sample_pnr_postselect
+        mock_eng_run = mock.MagicMock()
 
         with monkeypatch.context() as m:
-            m.setattr(sample, "_sample_sf", dummy_sample_sf)
-            samples = np.array(
-                sample.sample(
-                    A=adj,
-                    n_mean=1.0,
-                    samples=sample_number,
-                    backend_options={"backend": valid_backends, "threshold": True, "postselect": 1},
-                )
-            )
+            m.setattr(sf.Engine, "run", mock_eng_run)
+            sample.sample(A=adj, n_mean=1, threshold=False)
+            p_func = mock_eng_run.call_args[0][0]
 
-        samples_target = np.array([sample_pnr_postselect[0] for _ in range(sample_number)])
-        samples_target[samples_target >= 1] = 1
+        b = blackbird.dumps(sf.io.to_blackbird(p))
+        b_func = blackbird.dumps(sf.io.to_blackbird(p_func))
 
-        assert np.allclose(samples, samples_target)
+        assert b == b_func
```

Comment:
This is a little out of scope, but it would be nice if there was a method that compares two programs for "equal syntax" that did not depend on `blackbird`. Just a minor comment.

Another thing you could do here to be more explicit is to just check the last command to make sure it is the right one, in this example, something like:

```python
assert isinstance(p_func.circuit[-1].op, sf.ops.MeasureThreshold)
```

In this case, you do not need to even have `p`.

Reply:
Agree :+1:

Note that there is a related issue open for this: https://github.com/XanaduAI/strawberryfields/issues/135

Reply:
@josh146 good to know. For the purposes of this test though, it would be a little cleaner to check the last op, since it is the only thing that we are testing here, and not sure we need to compare the programs fully.

Reply:
Thanks @zzzeid this is awesome! 😁